File: dotnet/src/InternalUtilities/process/Runtime/ProcessMessageFactory.cs
// Copyright (c) Microsoft. All rights reserved.

using System.Collections.Generic;

namespace Microsoft.SemanticKernel.Process.Runtime;

/// <summary>
/// A factory class for creating <see cref="ProcessMessage"/> instances.
/// </summary>
internal static class ProcessMessageFactory
{
    /// <summary>
    /// Creates a new <see cref="ProcessMessage"/> instance from a <see cref="KernelProcessEdge"/> and a data object.
    /// </summary>
    /// <param name="edge">An instance of <see cref="KernelProcessEdge"/></param>
    /// <param name="sourceEventId">id of the source steps generating the event</param>
    /// <param name="data">A data object.</param>
    /// <param name="writtenToThread">Optional thread id where the event was written</param>
    /// <returns>An instance of <see cref="ProcessMessage"/></returns>
    internal static ProcessMessage CreateFromEdge(KernelProcessEdge edge, string sourceEventId, object? data, string? writtenToThread = null)
    {
        if (edge.OutputTarget is KernelProcessFunctionTarget functionTarget)
        {
            Dictionary<string, object?> parameterValue = [];
            if (!string.IsNullOrWhiteSpace(functionTarget.ParameterName))
            {
                parameterValue.Add(functionTarget.ParameterName!, data);
            }

            ProcessMessage newMessage = new(edge.SourceStepId, functionTarget.StepId, functionTarget.FunctionName, parameterValue)
            {
                SourceEventId = sourceEventId,
                TargetEventId = functionTarget.TargetEventId,
                TargetEventData = data,
                GroupId = edge.GroupId,
                writtenToThread = writtenToThread
            };

            return newMessage;
        }
        else if (edge.OutputTarget is KernelProcessAgentInvokeTarget agentTarget)
        {
            return new ProcessMessage(sourceEventId, agentTarget.StepId, "", [])
            {
                SourceEventId = sourceEventId,
                TargetEventId = null,
                TargetEventData = data,
                GroupId = edge.GroupId,
                writtenToThread = writtenToThread,
                MessagesInEval = agentTarget.MessagesInEval,
                InputEvals = agentTarget.InputEvals,
                ThreadEval = agentTarget.ThreadEval
            };
        }

        throw new KernelException($"Unsupported target type: {edge.OutputTarget.GetType().Name}");
    }
}

File: python/semantic_kernel/processes/process_message_factory.py
# Copyright (c) Microsoft. All rights reserved.

from typing import Any

from semantic_kernel.processes.kernel_process.kernel_process_edge import KernelProcessEdge
from semantic_kernel.processes.process_message import ProcessMessage
from semantic_kernel.utils.feature_stage_decorator import experimental


@experimental
class ProcessMessageFactory:
    """Factory class for creating ProcessMessage instances."""

    @staticmethod
    def create_from_edge(edge: KernelProcessEdge, data: Any) -> ProcessMessage:
        """Creates a new ProcessMessage from a KernelProcessEdge."""
        target = edge.output_target
        parameter_value: dict[str, Any] = {}
        if target.parameter_name is not None:
            parameter_value[target.parameter_name] = data

        return ProcessMessage(
            source_id=edge.source_step_id,
            destination_id=target.step_id,
            function_name=target.function_name,
            values=parameter_value,
            target_event_id=target.target_event_id,
            target_event_data=data,
        )
